src/lib/ndpi_main.c

```diff
@@ -5037,6 +5045,47 @@ void ndpi_connection_tracking(struct ndpi_detection_module_struct *ndpi_str,
 
 /* ************************************************ */
 
+static u_int32_t check_ndpi_subprotocols(struct ndpi_detection_module_struct * const ndpi_str,
+                                         struct ndpi_flow_struct * const flow,
+                                         NDPI_SELECTION_BITMASK_PROTOCOL_SIZE const ndpi_selection_packet,
+                                         NDPI_PROTOCOL_BITMASK detection_bitmask,
+                                         u_int16_t detected_protocol)
+{
+  u_int32_t num_calls = 0;
+
+  if (detected_protocol == NDPI_PROTOCOL_UNKNOWN)
+  {
+    return num_calls;
+  }
+
+  for (u_int32_t a = 0; a < ndpi_str->proto_defaults[detected_protocol].subprotocol_count; a++)
+  {
+    u_int16_t subproto_id = ndpi_str->proto_defaults[detected_protocol].subprotocols[a];
+    if (subproto_id == (uint16_t)NDPI_PROTOCOL_MATCHED_BY_CONTENT ||
+        subproto_id == flow->detected_protocol_stack[0] ||
+        subproto_id == flow->detected_protocol_stack[1])
+    {
+      continue;
+    }
+
+    u_int16_t subproto_index = ndpi_str->proto_defaults[subproto_id].protoIdx;
+    if ((ndpi_str->callback_buffer[subproto_index].ndpi_selection_bitmask & ndpi_selection_packet) ==
+         ndpi_str->callback_buffer[subproto_index].ndpi_selection_bitmask &&
+        NDPI_BITMASK_COMPARE(flow->excluded_protocol_bitmask,
+                             ndpi_str->callback_buffer[subproto_index].excluded_protocol_bitmask) == 0 &&
+        NDPI_BITMASK_COMPARE(ndpi_str->callback_buffer[subproto_index].detection_bitmask,
+                             detection_bitmask) != 0)
+    {
+      ndpi_str->callback_buffer[subproto_index].func(ndpi_str, flow);
+      num_calls++;
+    }
```

Comment:
Sorry but I don't get it: don't you have change the logic from "first-come-first-serve" to "the last match win"? Isn't that some kind of "dissector hierarchy" anyway?

But the main point I don't understand is: why a protocol hierarchy is fine with the "standard" classification but not for the "sub-protocol" classification (i.e. mainly/only the HTTP sub-classification)? Do you have any examples triggering any issue?

As usual, my main concern is performance: `check_ndpi_detection_func()` is already one of the hottest functions; I'd like to avoid calling more callbacks, if not necessary. Obviously, not problem from me if we need to...

Reply:
Thanks, and I agree with your general description.
But I don't get this point:
* with the previous code (with the break), the final classification was determined by the first match (i.e. the first matching protocol of the sub list)
*  without the break, the final classification is determined by the last match

I don't know which solution is the best one; but it seems to me that the classification results still depends on the sub-protocols list order.
Do you agree with that?

Reply:
>     * with the previous code (with the break), the final classification was determined by the first match (i.e. the first matching protocol of the sub list)

Exactly.

>     * without the break, the final classification is determined by the last match

Not necessarily. If a protocol sets a classification, others are still possible to overwrite that as long as they did not exclude themselves. That way the decision for a classification is delegated to the dissector instead the core.

Reply:
Ok, I am not entirely convinced, but this patch is an improvements compare to the present code.
Let's merge it and we will face any fallback in the future (if any)
Thanks